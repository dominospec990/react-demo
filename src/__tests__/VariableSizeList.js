import React from 'react';
import ReactTestRenderer from 'react-test-renderer';
import { VariableSizeList } from '..';

const findScrollContainer = rendered => rendered.root.children[0].children[0];

describe('VariableSizeList', () => {
  let cellRenderer, defaultProps;

  beforeEach(() => {
    jest.useFakeTimers();

    cellRenderer = jest.fn(({ style, ...rest }) => (
      <div style={style}>{JSON.stringify(rest, null, 2)}</div>
    ));
    defaultProps = {
      cellSize: index => 25 + index,
      children: cellRenderer,
      count: 20,
      estimatedCellSize: 25,
      height: 100,
      width: 50,
    };
  });

  // Much of the shared List functionality is already tested by FixedSizeList tests.
  // This test covers functionality that is unique to VariableSizeList.

  it('changing cellSize does not impact the rendered items', () => {
    const onItemsRendered = jest.fn();
    const rendered = ReactTestRenderer.create(
      <VariableSizeList {...defaultProps} onItemsRendered={onItemsRendered} />
    );
    rendered.update(
      <VariableSizeList
        {...defaultProps}
        cellSize={index => 50}
        onItemsRendered={onItemsRendered}
      />
    );
    expect(onItemsRendered.mock.calls).toMatchSnapshot();
  });

  describe('estimatedCellSize', () => {
    it('should estimate an initial scrollable height based on this estimation', () => {
      const rendered = ReactTestRenderer.create(
        <VariableSizeList {...defaultProps} estimatedCellSize={50} />
      );
      const scrollContainer = findScrollContainer(rendered);
      expect(scrollContainer.props.style).toMatchSnapshot();
    });

    it('should udpate the scrollable height as more cells are measured', () => {
      const rendered = ReactTestRenderer.create(
        <VariableSizeList {...defaultProps} estimatedCellSize={50} />
      );
      rendered.getInstance().scrollToItem(19);
      const scrollContainer = findScrollContainer(rendered);
      expect(scrollContainer.props.style).toMatchSnapshot();
    });
  });

  describe('scrollToItem method', () => {
    it('should scroll to the correct item for align = "auto"', () => {
      const onItemsRendered = jest.fn();
      const rendered = ReactTestRenderer.create(
        <VariableSizeList {...defaultProps} onItemsRendered={onItemsRendered} />
      );
      // Scroll down enough to show item 10 at the bottom.
      rendered.getInstance().scrollToItem(10, 'auto');
      // No need to scroll again; item 9 is already visible.
      // Overscan indices will change though, since direction changes.
      rendered.getInstance().scrollToItem(9, 'auto');
      // Scroll up enough to show item 2 at the top.
      rendered.getInstance().scrollToItem(2, 'auto');
      expect(onItemsRendered.mock.calls).toMatchSnapshot();
    });

    it('should scroll to the correct item for align = "start"', () => {
      const onItemsRendered = jest.fn();
      const rendered = ReactTestRenderer.create(
        <VariableSizeList {...defaultProps} onItemsRendered={onItemsRendered} />
      );
      // Scroll down enough to show item 10 at the top.
      rendered.getInstance().scrollToItem(10, 'start');
      // Scroll back up so that item 9 is at the top.
      // Overscroll direction wil change too.
      rendered.getInstance().scrollToItem(9, 'start');
      // Item 19 can't align at the top because there aren't enough items.
      // Scroll down as far as possible though.
      // Overscroll direction wil change again.
      rendered.getInstance().scrollToItem(19, 'start');
      expect(onItemsRendered.mock.calls).toMatchSnapshot();
    });

    it('should scroll to the correct item for align = "end"', () => {
      const onItemsRendered = jest.fn();
      const rendered = ReactTestRenderer.create(
        <VariableSizeList {...defaultProps} onItemsRendered={onItemsRendered} />
      );
      // Scroll down enough to show item 10 at the bottom.
      rendered.getInstance().scrollToItem(10, 'end');
      // Scroll back up so that item 9 is at the bottom.
      // Overscroll direction wil change too.
      rendered.getInstance().scrollToItem(9, 'end');
      // Item 1 can't align at the bottom because it's too close to the beginning.
      // Scroll up as far as possible though.
      // Overscroll direction wil change again.
      rendered.getInstance().scrollToItem(1, 'end');
      expect(onItemsRendered.mock.calls).toMatchSnapshot();
    });

    it('should scroll to the correct item for align = "center"', () => {
      const onItemsRendered = jest.fn();
      const rendered = ReactTestRenderer.create(
        <VariableSizeList {...defaultProps} onItemsRendered={onItemsRendered} />
      );
      // Scroll down enough to show item 10 in the middle.
      rendered.getInstance().scrollToItem(10, 'center');
      // Scroll back up so that item 9 is in the middle.
      // Overscroll direction wil change too.
      rendered.getInstance().scrollToItem(9, 'center');
      // Item 1 can't align in the middle because it's too close to the beginning.
      // Scroll up as far as possible though.
      // Overscroll direction wil change again.
      rendered.getInstance().scrollToItem(1, 'center');
      expect(onItemsRendered.mock.calls).toMatchSnapshot();
      // Item 19 can't align in the middle because it's too close to the end.
      // Scroll down as far as possible though.
      // Overscroll direction wil change again.
      rendered.getInstance().scrollToItem(19, 'center');
      expect(onItemsRendered.mock.calls).toMatchSnapshot();
    });
  });

  describe('resetAfterIndex method', () => {
    it('should reset cached sizes after the specified index', () => {
      // This impacts total height estimate.
      // Also impacts the size of cells rendered.
    });
  });

  describe('props validation', () => {
    beforeEach(() => spyOn(console, 'error'));

    it('should fail if non-function cellSize is provided', () => {
      expect(() =>
        ReactTestRenderer.create(
          <VariableSizeList {...defaultProps} cellSize={123} />
        )
      ).toThrow(
        'An invalid "cellSize" prop has been specified. ' +
          'Value should be a function. "number" was specified.'
      );
    });
  });
});
