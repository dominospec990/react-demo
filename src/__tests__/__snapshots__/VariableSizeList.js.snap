// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VariableSizeList onScroll scrolling should report partial items correctly in onItemsRendered 1`] = `
Array [
  Array [
    Object {
      "overscanStartIndex": 0,
      "overscanStopIndex": 6,
      "visibleStartIndex": 0,
      "visibleStopIndex": 4,
    },
  ],
  Array [
    Object {
      "overscanStartIndex": 1,
      "overscanStopIndex": 8,
      "visibleStartIndex": 2,
      "visibleStopIndex": 6,
    },
  ],
  Array [
    Object {
      "overscanStartIndex": 2,
      "overscanStopIndex": 8,
      "visibleStartIndex": 3,
      "visibleStopIndex": 6,
    },
  ],
  Array [
    Object {
      "overscanStartIndex": 0,
      "overscanStopIndex": 7,
      "visibleStartIndex": 2,
      "visibleStopIndex": 6,
    },
  ],
  Array [
    Object {
      "overscanStartIndex": 94,
      "overscanStopIndex": 99,
      "visibleStartIndex": 95,
      "visibleStopIndex": 99,
    },
  ],
  Array [
    Object {
      "overscanStartIndex": 95,
      "overscanStopIndex": 99,
      "visibleStartIndex": 96,
      "visibleStopIndex": 99,
    },
  ],
  Array [
    Object {
      "overscanStartIndex": 93,
      "overscanStopIndex": 99,
      "visibleStartIndex": 95,
      "visibleStopIndex": 99,
    },
  ],
]
`;

exports[`VariableSizeList scrollToItem method scroll with align = "auto" should work with partially-visible items 1`] = `
Array [
  Array [
    Object {
      "overscanStartIndex": 0,
      "overscanStopIndex": 5,
      "visibleStartIndex": 0,
      "visibleStopIndex": 3,
    },
  ],
  Array [
    Object {
      "overscanStartIndex": 5,
      "overscanStopIndex": 12,
      "visibleStartIndex": 7,
      "visibleStopIndex": 10,
    },
  ],
  Array [
    Object {
      "overscanStartIndex": 94,
      "overscanStopIndex": 99,
      "visibleStartIndex": 96,
      "visibleStopIndex": 99,
    },
  ],
  Array [
    Object {
      "overscanStartIndex": 0,
      "overscanStopIndex": 6,
      "visibleStartIndex": 1,
      "visibleStopIndex": 4,
    },
  ],
  Array [
    Object {
      "overscanStartIndex": 0,
      "overscanStopIndex": 5,
      "visibleStartIndex": 0,
      "visibleStopIndex": 3,
    },
  ],
]
`;

exports[`VariableSizeList scrollToItem method should not set invalid offsets when the list contains few items 1`] = `[Function]`;

exports[`VariableSizeList scrollToItem method should scroll to the correct item for align = "auto" 1`] = `
Array [
  Array [
    Object {
      "overscanStartIndex": 0,
      "overscanStopIndex": 5,
      "visibleStartIndex": 0,
      "visibleStopIndex": 3,
    },
  ],
  Array [
    Object {
      "overscanStartIndex": 6,
      "overscanStopIndex": 12,
      "visibleStartIndex": 8,
      "visibleStopIndex": 10,
    },
  ],
  Array [
    Object {
      "overscanStartIndex": 0,
      "overscanStopIndex": 7,
      "visibleStartIndex": 2,
      "visibleStopIndex": 5,
    },
  ],
]
`;

exports[`VariableSizeList scrollToItem method should scroll to the correct item for align = "center" 1`] = `
Array [
  Array [
    Object {
      "overscanStartIndex": 0,
      "overscanStopIndex": 5,
      "visibleStartIndex": 0,
      "visibleStopIndex": 3,
    },
  ],
  Array [
    Object {
      "overscanStartIndex": 7,
      "overscanStopIndex": 13,
      "visibleStartIndex": 9,
      "visibleStopIndex": 11,
    },
  ],
  Array [
    Object {
      "overscanStartIndex": 6,
      "overscanStopIndex": 12,
      "visibleStartIndex": 8,
      "visibleStopIndex": 10,
    },
  ],
  Array [
    Object {
      "overscanStartIndex": 0,
      "overscanStopIndex": 6,
      "visibleStartIndex": 0,
      "visibleStopIndex": 4,
    },
  ],
]
`;

exports[`VariableSizeList scrollToItem method should scroll to the correct item for align = "center" 2`] = `
Array [
  Array [
    Object {
      "overscanStartIndex": 0,
      "overscanStopIndex": 5,
      "visibleStartIndex": 0,
      "visibleStopIndex": 3,
    },
  ],
  Array [
    Object {
      "overscanStartIndex": 7,
      "overscanStopIndex": 13,
      "visibleStartIndex": 9,
      "visibleStopIndex": 11,
    },
  ],
  Array [
    Object {
      "overscanStartIndex": 6,
      "overscanStopIndex": 12,
      "visibleStartIndex": 8,
      "visibleStopIndex": 10,
    },
  ],
  Array [
    Object {
      "overscanStartIndex": 0,
      "overscanStopIndex": 6,
      "visibleStartIndex": 0,
      "visibleStopIndex": 4,
    },
  ],
  Array [
    Object {
      "overscanStartIndex": 15,
      "overscanStopIndex": 19,
      "visibleStartIndex": 17,
      "visibleStopIndex": 19,
    },
  ],
]
`;

exports[`VariableSizeList scrollToItem method should scroll to the correct item for align = "end" 1`] = `
Array [
  Array [
    Object {
      "overscanStartIndex": 0,
      "overscanStopIndex": 5,
      "visibleStartIndex": 0,
      "visibleStopIndex": 3,
    },
  ],
  Array [
    Object {
      "overscanStartIndex": 6,
      "overscanStopIndex": 12,
      "visibleStartIndex": 8,
      "visibleStopIndex": 10,
    },
  ],
  Array [
    Object {
      "overscanStartIndex": 4,
      "overscanStopIndex": 11,
      "visibleStartIndex": 6,
      "visibleStopIndex": 9,
    },
  ],
  Array [
    Object {
      "overscanStartIndex": 0,
      "overscanStopIndex": 5,
      "visibleStartIndex": 0,
      "visibleStopIndex": 3,
    },
  ],
]
`;

exports[`VariableSizeList scrollToItem method should scroll to the correct item for align = "smart" 1`] = `
Array [
  Array [
    Object {
      "overscanStartIndex": 0,
      "overscanStopIndex": 5,
      "visibleStartIndex": 0,
      "visibleStopIndex": 3,
    },
  ],
  Array [
    Object {
      "overscanStartIndex": 7,
      "overscanStopIndex": 13,
      "visibleStartIndex": 9,
      "visibleStopIndex": 11,
    },
  ],
  Array [
    Object {
      "overscanStartIndex": 4,
      "overscanStopIndex": 11,
      "visibleStartIndex": 6,
      "visibleStopIndex": 9,
    },
  ],
  Array [
    Object {
      "overscanStartIndex": 0,
      "overscanStopIndex": 6,
      "visibleStartIndex": 0,
      "visibleStopIndex": 4,
    },
  ],
]
`;

exports[`VariableSizeList scrollToItem method should scroll to the correct item for align = "start" 1`] = `
Array [
  Array [
    Object {
      "overscanStartIndex": 0,
      "overscanStopIndex": 5,
      "visibleStartIndex": 0,
      "visibleStopIndex": 3,
    },
  ],
  Array [
    Object {
      "overscanStartIndex": 8,
      "overscanStopIndex": 14,
      "visibleStartIndex": 10,
      "visibleStopIndex": 12,
    },
  ],
  Array [
    Object {
      "overscanStartIndex": 7,
      "overscanStopIndex": 13,
      "visibleStartIndex": 9,
      "visibleStopIndex": 11,
    },
  ],
  Array [
    Object {
      "overscanStartIndex": 15,
      "overscanStopIndex": 19,
      "visibleStartIndex": 17,
      "visibleStopIndex": 19,
    },
  ],
]
`;
